pennylane/plugins/tests/conftest.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains shared fixtures for the device tests."""
+import os
+import pytest
+import numpy as np
+import pennylane as qml
+
+# ==========================================================
+# pytest fixtures
+
+# seed for random functions
+np.random.seed(42)
+# Tolerance for analytic tests
+TOL = 1e-6
+# Tolerance for non-analytic tests
+TOL_STOCHASTIC = 0.05
+# Number of shots to call the devices with
+N_SHOTS = 10000
+# List of all devices that are included in PennyLane
+LIST_CORE_DEVICES = {"default.qubit", "default.qubit.tf"}
+
+
+@pytest.fixture(scope="function")
+def tol():

```

Comment:
Minor suggestion, but pytest allows for fixture injection. So for instance, you could do the following:

```python
@pytest.fixture(scope="function")
def tol(analytic):
    if analytic:
        return {"atol": float(os.environ.get("TOL", TOL)), "rtol": 0}
    return {"atol": TOL_STOCHASTIC, "rtol": 0}
```

where the `analytic` fixture is _not yet defined_.

When running the test, you simply declare the value of the analytic fixture using `@pytest.mark.parametrize`:

```python
@pytest.mark.parametrize(analytic, [True]):
def test_something(tol):
    ...
    assert np.allclose(result, expected, **tol)
```
the test doesn't even need to have `anlaytic` in the signature, it is injected directly into any dependent fixtures (in this case, `tol`).

You can also run the test for multiple values as usual:


```python
@pytest.mark.parametrize(analytic, [True]):
def test_something(analytic, tol):
    dev = qml.device("dsfasd", analytic=analytic)
    assert np.allclose(result, expected, **tol)
```

Reply:
We already use this pattern in the interface tests (fixtures have implicit dependence on an undefined `interface` fixture, which is then defined at the module/class/function level using `parametrize`). The main simplification of this approach is
* simplification of test signatures
* simplification of fixtures (no need to make them factories).

This comes at the expense that developers need to remember that `tol` requires that `analytic` be defined, even for a single value.

Reply:
I really like it (and it makes me wonder if I chose the right design in the first place), but "analytic" is not really a fixture chosen in a test, but a property of the device which is only created in the test itself...

I don't see how I can make use of fixture injection here?

Reply:
🤔